### PR TITLE
add compare to crypto system buf

### PR DESCRIPTION
**Pull request by zippy**
## PR summary

adds compare function to crypto-system Buffer trait, and copies over the implementation for sodium secure buffer and does a bad implementation for u8 buffer.

### CODE REVIEW

File: crates/crypto_api/src/buffer/buffer_vec_u8.rs [hunk 1]
```diff
@@ -19,4 +19,21 @@ impl Buffer for Vec<u8> {
...
+        let l = self.len();
+        if l != b.len() {
+            return 1;
+        }
+        let a = self.read_lock();
+        let b = b.read_lock();
+        let mut i = 0;
+        while i < l {
+            if a[i] != b[i] {
+                return 1;
```
Comment (neonphog) [inline crates/crypto_api/src/buffer/buffer_vec_u8.rs]: We actually need this to return -1 if a < b when interpreted as little endian per https://download.libsodium.org/doc/helpers#comparing-large-numbers.

We could potentially use https://docs.rs/libc/0.2.63/libc/fn.memcmp.html (though I can't remember, we might need to reverse the result)

File: crates/crypto_api/src/buffer/buffer_vec_u8.rs [hunk 2]
```diff
@@ -19,4 +19,21 @@ impl Buffer for Vec<u8> {
     fn set_no_access(&self) {}
     fn set_readable(&self) {}
     fn set_writable(&self) {}
+
+    fn compare(&mut self, b: &mut Box<dyn Buffer>) -> i32 {
```
Comment (neonphog) [inline crates/crypto_api/src/buffer/buffer_vec_u8.rs]: what do you think about doing this as a provided implementation in the Trait, like https://github.com/holochain/lib3h/blob/develop/crates/crypto_api/src/buffer/mod.rs#L59-L62 ?

Comment (zippy) [reply crates/crypto_api/src/buffer/buffer_vec_u8.rs]: done

File: crates/crypto_api/src/buffer/buffer_vec_u8.rs (hunk 1 above)
Comment (zippy) [reply crates/crypto_api/src/buffer/buffer_vec_u8.rs]: code now compares values, but probably doesn't do the right thing on different length buffers, but at lease I added a test.  @neonphog can you confirm what the appropriate expected behavior of compare should be?

**Comment by zippy** [comment]
@neonphog added yet another test to make things clear and I'm quite mystified by the sodium compare algorithm  data seems to show that:

[45, 0] < [0, 1]  and
[45] > [0, 1]

that seems weird to me.

**Comment by ddd-mtl** [comment]
found the issue. It compares right to left, starting from a's size.

File: crates/sodium/src/crypto_system/secure_buffer.rs [hunk 3]
```diff
@@ -133,6 +133,12 @@ impl Buffer for SecureBuffer {
         }
         *self.p.borrow_mut() = ProtectState::ReadWrite;
     }
+
+    fn compare(&mut self, b: &mut Box<dyn Buffer>) -> i32 {
+        let mut a = self.write_lock();
+        let mut b = b.write_lock();
+        unsafe { rust_sodium_sys::sodium_compare(raw_ptr_char!(a), raw_ptr_char!(b), a.len()) }
+    }
```
Comment (neonphog) [inline crates/sodium/src/crypto_system/secure_buffer.rs]: I think this should all be immutable, no?

```rust
    fn compare(&self, b: &Box<dyn Buffer>) -> i32 {
        let mut a = self.read_lock();
        let mut b = b.read_lock();
        unsafe { rust_sodium_sys::sodium_compare(raw_ptr_char_immut!(a), raw_ptr_char_immut!(b), a.len()) }
    }
```

File: crates/crypto_api/src/buffer/mod.rs [hunk 4]
```diff
@@ -73,6 +73,30 @@ pub trait Buffer: Send + std::fmt::Debug + Deref<Target = [u8]> + DerefMut<Targe
...
         Ok(())
     }
+
+    /// compare this buffer to another buffer
+    /// Return :
+    /// | if a > b; return 1
+    /// | if a < b; return -1
+    /// | if a == b; return 0
+    #[allow(clippy::borrowed_box)]
+    fn compare(&mut self, b: &mut Box<dyn Buffer>) -> i32 {
```
Comment (neonphog) [inline crates/crypto_api/src/buffer/mod.rs]: see other comment, this should be immutable
```suggestion
    fn compare(&self, b: &Box<dyn Buffer>) -> i32 {
```

**Pull request merged**